### PR TITLE
Omit scm-username annotation from the PAT secret

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Ignore the `che.eclipse.org/scm-username` annotation when fetching token from a PAT secret. Instead fetch the username from the `isValid(<token>)` request. 

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/22344

### How to test this PR?

1. Deploy che using the PR image: `quay.io/ivinokur/che-server:che-22344`
2. Create a PAT of any of the supported scm providers:
 - Go to dashboard -> user-preferences -> Personal Access Tokens tab and create a token but enter a random string in the `username` input.
 - Create a [PAT secret in the user namespace manually](https://www.eclipse.org/che/docs/stable/end-user-guide/using-a-git-provider-access-token/) but either omit the `che.eclipse.org/scm-username` annotation or use a random string value
 3. Create a workspace from a private repository of the used scm provider.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java [hunk 1]
```diff
@@ -49,6 +51,19 @@ PersonalAccessToken fetchPersonalAccessToken(Subject cheUser, String scmServerUr
...
+   * Checks whether the provided personal access token is valid by fetching user info from the scm
+   * provider. Also checks whether the token has expected scope of permissions if the provider API
+   * supports such request.
+   *
+   * @return - Optional with a pair of boolean value and scm username. The boolean value is true if
+   *     the token has expected scope of permissions, false if the token scopes does not match the
+   *     expected ones. Empty optional if {@link PersonalAccessTokenFetcher} is not able to confirm
+   *     or deny that token is valid.
+   */
+  Optional<Pair<Boolean, String>> isValid(PersonalAccessTokenParams params);
```
Comment (amisevsk) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java:68]: I'm a little confused by `isValid` being updated to return the username in addition to whether the token is valid. Perhaps these should be separated into multiple functions?

**Review by amisevsk — COMMENTED** [review]
I'm confused by the three versions of `isValid` in this PR -- one returns a bool, one returns a string, and one returns `pair<bool, string>`

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java (hunk 1 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java:68]: The `isValid()` function verifies the token by requesting an scm user. The idea is to return the username in order not to request the username on demand. This reduces the number of API calls.

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java [hunk 2]
```diff
@@ -55,6 +57,7 @@ public PersonalAccessToken fetchPersonalAccessToken(Subject cheUser, String scmS
    * Iterate over the Set<PersonalAccessTokenFetcher> declared in container and sequentially invoke
    * {@link PersonalAccessTokenFetcher#isValid(PersonalAccessToken)} method.
    *
+   * @deprecated use {@link #isValid(PersonalAccessTokenParams)} instead.
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java:60]: yes, added

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java [hunk 3]
```diff
@@ -71,4 +75,21 @@ public boolean isValid(PersonalAccessToken personalAccessToken)
...
         personalAccessToken.getScmProviderUrl());
   }
+
+  /**
+   * Iterate over the Set<PersonalAccessTokenFetcher> declared in container and sequentially invoke
+   * {@link PersonalAccessTokenFetcher#isValid(PersonalAccessTokenParams)} method. If any of the
+   * fetchers return an scm username, return it. Otherwise, return null.
+   */
+  @Nullable
+  public String isValid(PersonalAccessTokenParams params)
```
Comment (tolusha) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java:85]: I am wondering if we can rename these `isValid` functions on `getUserName`

Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java]: done

**Comment by tolusha** [comment]
Tested for Azure DevOps

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java (hunk 1 above)
Comment (amisevsk) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java:68]: Maybe `getUser()` is a better name?

Comment (amisevsk) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java:68]: Ah I see the discussion below :)

**Comment by artaleks9** [comment]
@vinokurig 
Verified for GitLab.com - works properly.
But for GitHub.com - there is a problem:
- The `github-with-pat-setup-flow` test crashes
with PAT configured for private GitHub repo, factory resolver doesn't return 200..
- I tried to manually set up PAT for GitHub.com and run factory - and I get an error..

**Comment by vinokurig** [comment]
Rolled back the PAT scopes validation for GitHub and GitLab, created a separate issue for that: https://github.com/eclipse/che/issues/22381

File: wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketServerURLParser.java [hunk 4]
```diff
@@ -114,9 +114,9 @@ private boolean isApiRequestRelevant(String repositoryUrl) {
               new BitbucketServerOAuthAuthenticator("", "", "", ""),
               oAuthAPI,
               "");
-      // If the token request catches the unauthorised error, it means that the provided url
+      // If the user request catches the unauthorised error, it means that the provided url
       // belongs to Bitbucket.
-      bitbucketServerApiClient.getPersonalAccessToken("", 0L);
+      bitbucketServerApiClient.getUser(null);
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketServerURLParser.java:119]: why do we pass null as an argument?

Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketServerURLParser.java:119]: If `null` is passed the client fetches OAuth token:
https://github.com/eclipse-che/che-server/blob/1657ebbaa9582bda79df30dc31f8003fadcd0654/wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/server/BitbucketServerApiClient.java#L30-L38

**Comment by artaleks9** [comment]
@vinokurig 
It seems, there is a some problem with deploy Che using the last `quay.io/eclipse/che-server:pr-533` image of PR:
```
[22:31:41] → Failed to start a pod, reason: Error, exitCode: 1
[22:31:41] Eclipse Che Server pod bootstrap [failed]
[22:31:41] → Failed to start a pod, reason: Error, exitCode: 1
[22:31:41] Wait for Eclipse Che ready [failed]
[22:31:41] → Failed to start a pod, reason: Error, exitCode: 1
    Error: Command server:deploy failed with the error: Failed to start a pod,
```
I tried to install Che manually, but got the error with start `che-server` pod, here is the `che-pod` log:
```
Using custom assembly from /home/user/eclipse-che
Found a custom cert. Adding it to java trust store /home/user/cacerts based on /usr/lib/jvm/jre/lib/security/cacerts
Exception in thread "main" java.lang.Error: java.io.FileNotFoundException: /usr/lib/jvm/java-11-openjdk-11.0.20.0.8-2.el8.x86_64/lib/tzdb.dat (No such file or directory)
	at java.base/sun.util.calendar.ZoneInfoFile$1.run(ZoneInfoFile.java:261)
	at java.base/sun.util.calendar.ZoneInfoFile$1.run(ZoneInfoFile.java:251)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/sun.util.calendar.ZoneInfoFile.<clinit>(ZoneInfoFile.java:251)
	at java.base/sun.util.calendar.ZoneInfo.getTimeZone(ZoneInfo.java:588)
	at java.base/java.util.TimeZone.getTimeZone(TimeZone.java:577)
	at java.base/java.util.TimeZone.setDefaultZone(TimeZone.java:682)
	at java.base/java.util.TimeZone.getDefaultRef(TimeZone.java:653)
	at java.base/java.util.Date.normalize(Date.java:1198)
	at java.base/java.util.Date.toString(Date.java:1031)
	at java.base/sun.security.tools.keytool.Main.printX509Cert(Main.java:3537)
	at java.base/sun.security.tools.keytool.Main.addTrustedCert(Main.java:3317)
	at java.base/sun.security.tools.keytool.Main.doCommands(Main.java:1222)
	at java.base/sun.security.tools.keytool.Main.run(Main.java:421)
	at java.base/sun.security.tools.keytool.Main.main(Main.java:414)
Caused by: java.io.FileNotFoundException: /usr/lib/jvm/java-11-openjdk-11.0.20.0.8-2.el8.x86_64/lib/tzdb.dat (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/sun.util.calendar.ZoneInfoFile$1.run(ZoneInfoFile.java:255)
	... 14 more
```

**Comment by vinokurig** [comment]
@artaleks9 Please use `quay.io/ivinokur/che-server:che-22344` image, built by myself

**Comment by vinokurig** [comment]
@artaleks9 
> It seems, there is a some problem with deploy Che using the last quay.io/eclipse/che-server:pr-533 image of PR

Fixed

**Comment by artaleks9** [comment]
@vinokurig 
I tested this PR with your `quay.io/ivinokur/che-server:che-22344`on GitHub.com and GitLab.com.
The functionality properly works on both providers.
One thing for GitHub - a git provider user name is wrong, I'll create a separate issue about that. See screenshots:
pr-533-workspace-userdata-gitconfig-configmap

**Review by artaleks9 — APPROVED** [review]

**Comment by vinokurig** [comment]
@artaleks9 
> https://github.com/eclipse-che/che-server/pull/533#issuecomment-1649167881

Is it reproducible only with this PR or in the original latest as well?

**Comment by artaleks9** [comment]
@vinokurig 
In the original latest too, it's not related to this PR.

**Review by ibuziuk — COMMENTED** [review]
I'm not super happy with returning in `getScmUsername`null and passing null as parameters in`getUser(null)` `getUserSlug(null)` 
I believe we should avoid that as much as possible and fall back on Optional / Optional.empty() etc

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java [hunk 5]
```diff
@@ -71,4 +75,21 @@ public boolean isValid(PersonalAccessToken personalAccessToken)
...
+  @Nullable
+  public String getScmUsername(PersonalAccessTokenParams params)
+      throws UnknownScmProviderException, ScmUnauthorizedException, ScmCommunicationException {
+    for (PersonalAccessTokenFetcher fetcher : personalAccessTokenFetchers) {
+      Optional<Pair<Boolean, String>> isValid = fetcher.isValid(params);
+      if (isValid.isPresent() && isValid.get().first) {
+        return isValid.get().second;
+      }
+    }
+    return null;
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java:93]: @vinokurig can we return Optional here?

File: wsmaster/che-core-api-factory-bitbucket-server/src/test/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClientTest.java [hunk 6]
```diff
@@ -325,17 +331,17 @@ public void shouldGetOauth2Token()
...
+        get(urlEqualTo("/rest/api/1.0/users/ksmster"))
             .withHeader(HttpHeaders.AUTHORIZATION, equalTo("Bearer token"))
             .willReturn(
                 aResponse()
                     .withHeader("Content-Type", "application/json; charset=utf-8")
                     .withBodyFile("bitbucket/rest/api/1.0/users/ksmster/response.json")));
 
     // when
-    bitbucketServer.getUser("user", null);
+    bitbucketServer.getUser(Optional.empty());
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory-bitbucket-server/src/test/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClientTest.java:342]: @vinokurig would it be possible to have a `getUser` method without parameters?

Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket-server/src/test/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClientTest.java]: That's a good point, done

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java (hunk 5 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/ScmPersonalAccessTokenFetcher.java]: done

File: wsmaster/che-core-api-factory-gitlab/src/test/resources/__files/gitlab/rest/api/v4/user/PAT_info.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "id": 1,
+  "name": "gitlab",
+  "scopes": ["api", "write_repository"],
+  "expires_at": "2024-07-18",
+  "created_at": "2023-07-19T07:41:19.492Z"
+}
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory-gitlab/src/test/resources/__files/gitlab/rest/api/v4/user/PAT_info.json:7]: no new line

File: wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClient.java [hunk 8]
```diff
@@ -288,20 +300,24 @@ public BitbucketPersonalAccessToken createPersonalAccessTokens(
...
   }
 
   @Override
-  public BitbucketPersonalAccessToken getPersonalAccessToken(String userSlug, Long tokenId)
+  public BitbucketPersonalAccessToken getPersonalAccessToken(Long tokenId)
       throws ScmItemNotFoundException, ScmUnauthorizedException, ScmCommunicationException {
-    URI uri = serverUri.resolve("./rest/access-tokens/1.0/users/" + userSlug + "/" + tokenId);
+    URI uri =
+        serverUri.resolve(
+            "./rest/access-tokens/1.0/users/" + getUserSlug(Optional.empty()) + "/" + tokenId);
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClient.java:320]: can we have `getuserSlug()` method?

File: wsmaster/che-core-api-factory-gitlab/src/test/resources/__files/gitlab/rest/api/v4/user/PAT_info.json (hunk 7 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-gitlab/src/test/resources/__files/gitlab/rest/api/v4/user/PAT_info.json]: fixed

File: wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClient.java (hunk 8 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/HttpBitbucketServerApiClient.java]: done

**Review by ibuziuk — APPROVED** [review]
:+1:

**Pull request merged**